libmcount/internal.h

```diff
@@ -145,20 +144,24 @@ static inline void mcount_restore_arch_context(struct mcount_arch_context *ctx)
 
 #ifdef SINGLE_THREAD
 # define TLS
-# define get_thread_data()  &mtd
-# define check_thread_data(mtdp)  (mtdp->rstack == NULL)
+# define attr_tls_model
 #else
 # define TLS  __thread
-# define get_thread_data()  pthread_getspecific(mtd_key)
-# define check_thread_data(mtdp)  (mtdp == NULL)
+# define attr_tls_model __attribute__((tls_model("initial-exec")))
 #endif
 
-extern TLS struct mcount_thread_data mtd;
+extern TLS struct mcount_thread_data *_mtdp attr_tls_model;
+extern TLS bool mcount_recursion_marker attr_tls_model;
```

Comment:
You need to add explanation why you use a pointer to mtd rather than the mtd itself and why the recursion marker is separated from the mtd.